### ISSUE TITLE
Haskell executable not being compiled

### ISSUE THREAD

**Issue by shuheng-liu** [wsdjeg/SpaceVim#3129]
## Expected behavior, english is recommend
The current haskell source file should be compiled and executed.

## Environment Information

- OS: mac
- vim version: 8.1.1950
- neovim version: -
- SpaceVim version: 1.3.0-dev
- SpaceVim status: 

```
On branch master
Your branch is up to date with 'origin/master'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)

    bundle/

nothing added to commit but untracked files present (use "git add" to track)
```

## The reproduce ways from Vim starting (Required!)
1. run `vim temp.hs` in terminal
2. type in `main = putStrLn "Hello World"`
3. press <kbd>SPC l r</kbd> (`lang#haskell` layers has been added)

The program should be executed but I got 
```bash
[Compile] ghc -v0 --make test.hs -o /private/var/folders/jr/qgh_hd1945bb_g0k8fwy_gn00000gp/T/vFxxEN4/2
[Running] /private/var/folders/jr/qgh_hd1945bb_g0k8fwy_gn00000gp/T/vFxxEN4/2

--------------------
executing job failed: No such file or directory

[Done] exited with code=122 in 0.007302 seconds
```
![Screen Shot 2019-10-06 at 17 10 38](https://user-images.githubusercontent.com/22414322/66266895-b357f780-e85d-11e9-8b2a-43f1a4460c0a.jpg)

In my understanding, the first line `ghc -v0 --make test.hs -o <PATH>` is compiling an executable `2` in some directory SpaceVim created, and the second line is running the executable. However, when I check for the existence of the executable, it is not there. The directory has only a `0.js` file which says 
```javascript
module.exports = {
  settings: {
    listItemIndent: '1',
  },
  plugins: [
    require('remark-frontmatter'),
  ]
};
```
If I try to manually compile and run the executable (using exactly the same 2 commands above) , it works as expected. Considering the error message, I think there is something going wrong with the compiling part in SpaceVim.
![Screen Shot 2019-10-06 at 17 20 41](https://user-images.githubusercontent.com/22414322/66266896-b81cab80-e85d-11e9-8377-971b3e9fd38d.jpg)

## Output of the `:SPDebugInfo!`

<details><summary> SpaceVim debug information </summary>

### SpaceVim options :

```toml
  auto_disable_touchpad = 1
  autocomplete_method = 'neocomplete'
  autocomplete_parens = 1
  bootstrap_after = 'myspacevim#after'
  bootstrap_before = 'myspacevim#before'
  buffer_index_type = 4
  checkinstall = 1
  colorscheme = 'gruvbox'
  colorscheme_bg = 'dark'
  colorscheme_default = 'desert'
  commandline_prompt = '➭'
  custom_color_palette = []
  custom_plugins = []
  debug_level = 1
  default_indent = 2
  disabled_plugins = []
  enable_ale = 0
  enable_bepo_layout = 0
  enable_cursorcolumn = 0
  enable_cursorline = 1
  enable_debug = 0
  enable_googlesuggest = 0
  enable_guicolors = 1
  enable_insert_leader = 1
  enable_javacomplete2_py = 0
  enable_key_frequency = 0
  enable_language_specific_leader = 1
  enable_neocomplcache = 0
  enable_neomake = 1
  enable_os_fileformat_icon = 0
  enable_powerline_fonts = 1
  enable_statusline_bfpath = 0
  enable_statusline_mode = 0
  enable_statusline_tag = 1
  enable_tabline_filetype_icon = 1
  enable_tabline_ft_icon = 0
  enable_vimfiler_filetypeicon = 0
  enable_vimfiler_gitstatus = 0
  enable_vimfiler_welcome = 1
  enable_ycm = 0
  error_symbol = '✖'
  expand_tab = 1
  filemanager = 'vimfiler'
  filetree_direction = 'right'
  filetype_icons = {}
  force_global_config = 0
  gitcommit_issue_icon = ''
  gitcommit_pr_icon = ''
  github_username = ''
  guifont = ''
  hiddenfileinfo = 1
  home_files_number = 6
  hosts_url = 'https://raw.githubusercontent.com/racaljk/hosts/master/hosts'
  info_symbol = 'ⓘ'
  keep_server_alive = 1
  language = ''
  layer_lang_java_formatter = '/Users/liushuheng/.external_tools/google-java-format-1.7.jar'
  leader_guide_default_group_name = ''
  leader_guide_flatten = 1
  leader_guide_hspace = 5
  leader_guide_max_size = 0
  leader_guide_position = 'botright'
  leader_guide_run_map_on_popup = 1
  leader_guide_sort_horizontal = 0
  leader_guide_submode_mappings = {'<C-C>': 'win_close'}
  leader_guide_vertical = 0
  lint_on_save = 1
  lint_on_the_fly = 0
  max_column = 120
  plugin_bundle_dir = '/Users/liushuheng/.cache/vimfiles/'
  plugin_manager = 'dein'
  plugin_manager_processes = 16
  plugin_name = 'tagbar-markdown'
  project_rooter_automatically = 1
  project_rooter_patterns = ['.git/', '_darcs/', '.hg/', '.bzr/', '.svn/', 'package.json', '.SpaceVim.d/', '.clang', 'pom.xml']
  realtime_leader_guide = 1
  relativenumber = 1
  search_tools = ['rg', 'ag', 'pt', 'ack', 'grep', 'findstr']
  sidebar_direction = ''
  sidebar_width = 30
  smartcloseignoreft = ['tagbar', 'vimfiler', 'defx', 'SpaceVimRunner', 'SpaceVimREPL', 'SpaceVimQuickFix', 'HelpDescribe', 'VebuggerShell', 'VebuggerTerminal', 'SpaceVimTabsManager']
  smartcloseignorewin = ['__Tagbar__', 'vimfiler:default']
  snippet_engine = 'neosnippet'
  src_root = 'E:\sources\'
  statusline_inactive_separator = 'arrow'
  statusline_iseparator = 'nil'
  statusline_left_sections = ['winnr', 'filename', 'major mode', 'search count', 'syntax checking', 'minor mode lighters']
  statusline_right_sections = ['fileformat', 'cursorpos', 'percentage']
  statusline_separator = 'arrow'
  statusline_unicode_symbols = 1
  terminal_cursor_shape = 2
  update_retry_cnt = 3
  version = '1.3.0-dev'
  vim_help_language = 'en'
  vimcompatible = 0
  warning_symbol = '⚠'
  wildignore = '*/tmp/*,*.so,*.swp,*.zip,*.class,tags,*.jpg,*.ttf,*.TTF,*.png,*/target/*,.git,.svn,.hg,.DS_Store,*.svg'
  windows_index_type = 3
  windows_leader = 's'
  windows_smartclose = 'q'
```

### SpaceVim layers :

```toml
[[layers]]
  name="autocomplete"
  auto-completion-return-key-behavior='complete'
  auto-completion-tab-key-behavior='smart'
[[layers]]
  name="checkers"
[[layers]]
  name="format"
[[layers]]
  name="edit"
[[layers]]
  name="ui"
[[layers]]
  name="core"
[[layers]]
  name="core#banner"
[[layers]]
  name="core#statusline"
[[layers]]
  name="core#tabline"
[[layers]]
  name="shell"
[[layers]]
  name="lang#python"
[[layers]]
  name="lang#javascript"
  auto_fix=1
[[layers]]
  name="lang#c"
[[layers]]
  name="lang#html"
[[layers]]
  name="lang#haskell"
[[layers]]
  name="lang#java"
[[layers]]
  name="lang#sh"
[[layers]]
  name="tools#dash"
[[layers]]
  name="lang#markdown"
[[layers]]
  name="-l"
```

### SpaceVim Health checking :

SpaceVim clipboard support check report:
Checking +clipboard:
      SUCCEED!
SpaceVim environment check report:
Current progpath: vim(vim)
version: 801
OS: mac
[shell, shellcmdflag, shellslash]: ['/bin/zsh', '-c', 0]
SpaceVim lua support check report:
Checking +lua:
      SUCCEED!
SpaceVim python support check report:
Checking +python3:
      SUCCEED!
Checking +python:
      Failed : to support +python, Please install vim-gik, or build from sources.

### SpaceVim runtime log :

```log
[ SpaceVim ] : logger file  does not exists, only log for current process will be shown!
[ SpaceVim ] [17:04:58] [ Info ] Startup with argv: ['test.hs']
[ SpaceVim ] [17:04:58] [ Info ] find local conf: /Users/liushuheng/.SpaceVim.d/init.toml
[ SpaceVim ] [17:04:58] [ Info ] loadding cached local conf: /Users/liushuheng/.cache/SpaceVim/conf/_Users_liushuheng__SpaceVim_d_init_toml.json
[ SpaceVim ] [17:04:58] [ Info ] start to apply config [local]
[ SpaceVim ] [17:04:59] [ Info ] Start to find root for: /Users/liushuheng/test.hs
[ SpaceVim ] [17:04:59] [ Info ]         (package.json):/Users/liushuheng/package.json
[ SpaceVim ] [17:04:59] [ Info ] change to root:/Users/liushuheng
[ SpaceVim ] [17:04:59] [ Info ] change to root:/Users/liushuheng
[ SpaceVim ] [17:05:05] [ Info ] Start to find root for: /Users/liushuheng/__runner__
[ SpaceVim ] [17:05:05] [ Info ]         (package.json):/Users/liushuheng/package.json
[ SpaceVim ] [17:05:05] [ Info ] change to root:/Users/liushuheng
[ SpaceVim ] [17:05:05] [ Info ] change to root:/Users/liushuheng
[ SpaceVim ] [17:05:26] [ Info ] Start to find root for: /Users/liushuheng/SpaceVimLayers
[ SpaceVim ] [17:05:26] [ Info ]         (package.json):/Users/liushuheng/package.json
[ SpaceVim ] [17:05:26] [ Info ] change to root:/Users/liushuheng
[ SpaceVim ] [17:05:42] [ Info ] change to root:/Users/liushuheng
[ SpaceVim ] [17:05:52] [ Info ] Start to find root for: /var/folders/jr/qgh_hd1945bb_g0k8fwy_gn00000gp/T/vFxxEN4/3/issue_report.md
[ SpaceVim ] [17:05:52] [ Info ] change to root:/Users/liushuheng
```
</details>

## Screenshots

If you have any screenshots for this issue please upload here. BTW you can use https://asciinema.org/ for recording video  in terminal.

**Comment by wsdjeg** [comment]
It look strange, you run the same command, and it works well in command line? ok, I will check it. thanks for reporting.

EDIT， I can not reproduce it now.

**Comment by shuheng-liu** [comment]
So maybe it’s not a bug, but a specific issue on my part. I’d much appreciate any suggestions to fix it.

**Comment by shuheng-liu** [comment]
Also, I don't know if it's relevant; but when I run `:SPUpdate`, everything updates fine except that the `vim-node-rpc` fails. 
Edit: I fixed the `vim-node-rps` problem but the haskell compiling issue remains. I think it must be caused by vim ***not doing the compilation while printing a message that it did***. I have no clue why.

**Comment by wsdjeg** [comment]
the runner do not require vim-node-rpc. it only base on neovim job api.

**Comment by Unactived** [comment]
I have the same issue on a fresh install of SpaceVim on Vim (not NeoVim)
![image](https://user-images.githubusercontent.com/33038089/66456743-bbd34c80-ea6e-11e9-8284-f87f84fbefd3.png)
I looked up the tmp directory. It exists, but is completely empty.

**Comment by Unactived** [comment]
> the runner do not require vim-node-rpc. it only base on neovim job api.

Reading this, I decided to try this on a fresh install on SpaceVim but this time for Neovim

I get a different result, yet still not the expected one:
the job is started but seems to never end:
![image](https://user-images.githubusercontent.com/33038089/66505324-6da06680-eacb-11e9-965f-d1703577da80.png)
Again, I checked that the directory /tmp/nvimW64lMb existed (it does) but it doesn't contain 2 (of course). It contains a socket named `0`.

Looking again, the tmp directory for Vim also has a socket. It seems the issues are that the job isn't started on Vim but is on NeoVim, and doesn't end anyway.

**Comment by shuheng-liu** [comment]
> > the runner do not require vim-node-rpc. it only base on neovim job api.
> 
> Reading this, I decided to try this on a fresh install on SpaceVim but this time for Neovim
> 
> I get a different result, yet still not the expected one:
> the job is started but seems to never end:
> ![image](https://user-images.githubusercontent.com/33038089/66505324-6da06680-eacb-11e9-965f-d1703577da80.png)
> Again, I checked that the directory /tmp/nvimW64lMb existed (it does) but it doesn't contain 2 (of course). It contains a socket named `0`.
> 
> Looking again, the tmp directory for Vim also has a socket. It seems the issues are that the job isn't started on Vim but is on NeoVim, and doesn't end anyway.

The exact same thing happens to me when I try using `nvim` instead of `vim`. The job seems never finishing.

**Comment by Unactived** [comment]
(Neovim)
The bug comes from this function: [job.self.start](https://github.com/SpaceVim/SpaceVim/blob/master/autoload/SpaceVim/api/job.vim#L102)

In the function, `job` takes the value -1, meaning jobstart found the command 'not executable'. Which is weird, as the command is working fine from shell (or through :! which is equivalent).

One thing though, as job is `-1`, "Failed to start job..." should be displayed, but it insn't. My guess is that we're adding the error message to `self._message` but we don't seem to use this var afterwards.

Not really related, but is [`endi`](https://github.com/SpaceVim/SpaceVim/blob/master/autoload/SpaceVim/api/job.vim#L109) a typo of `endif` in this file?

I didn't check the "Vim version of the problem" though it should be the same, but actually less of a problem as it appears that in this case, we at least get the error message.

I'm afraid to not be really helpful as I have no experience with VimScript.

**Comment by wsdjeg** [comment]
thanks，I will look into it.



发自我的小米手机
在 FrenchMasterSword <notifications@github.com>，2019年10月12日 05:17写道：

(Neovim)
The bug comes from this function: job.self.start<https://github.com/SpaceVim/SpaceVim/blob/master/autoload/SpaceVim/api/job.vim#L102>

In the function, job takes the value -1, meaning jobstart found the command 'not executable'. Which is weird, as the command is working fine from shell (or through :! which is equivalent).

One thing though, as job is -1, "Failed to start job..." should be displayed, but it insn't. My guess is that we're adding the error message to self._message but we don't seem to use this var afterwards.

Not really related, but is endi<https://github.com/SpaceVim/SpaceVim/blob/master/autoload/SpaceVim/api/job.vim#L109> a typo of endif in this file?

I didn't checked the "Vim version of the problem" though it should be the same, but actually less of a problem as it appears that in this case, we at least get the error message.

I'm afraid to not be really helpful as I have no experience with VimScript.

**Comment by wsdjeg** [comment]
> Not really related, but is endi a typo of endif in this file?

It is a bug of this file, I will check it.

**Comment by Unactived** [comment]
Sorry to bring this up again.

The referenced commit only fixes the typo. Which doesn't fix the issue itself (when signaling this typo I had myself tried this change to see if it fixed it).

The issue was closed but the problem wasn't fixed.

**Comment by wsdjeg** [comment]
@Johnny-Wish I just fixed this issue in #3292 

![haskell_runner](https://user-images.githubusercontent.com/13142418/72685218-fd21d080-3b22-11ea-8c84-14cfef171cf8.png)
